core/lib/zksync_core/src/native_erc20_fetcher/mod.rs

```diff
@@ -0,0 +1,107 @@
+use std::{sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use metrics::atomics::AtomicU64;
+use tokio::{
+    sync::{watch, OnceCell},
+    task::JoinHandle,
+};
+use zksync_config::configs::native_erc20_fetcher::NativeErc20FetcherConfig;
+
+// TODO: this error type is also defined by the gasAdjuster module,
+//       we should consider moving it to a common place
+#[derive(thiserror::Error, Debug, Clone)]
+#[error(transparent)]
+pub struct Error(Arc<anyhow::Error>);
+
+impl From<anyhow::Error> for Error {
+    fn from(err: anyhow::Error) -> Self {
+        Self(Arc::new(err))
+    }
+}
+
+pub trait Erc20Fetcher: 'static + std::fmt::Debug + Send + Sync {
+    fn conversion_rate(&self) -> anyhow::Result<u64>;
+}
+
+pub(crate) struct NativeErc20FetcherSingleton {
```

Comment:
I'm not entirely sure I follow here, you can pass `None` to `GasAdjusterSingleton` when creating it, so you don't need to instantiate `Erc20Fetcher` to create `GasAdjusterSingleton`

Reply:
I meant the case when `NativeERC20Fetcher` component is provided, but no `GasAdjuster` instance is created.

Let's look at the examples:
- No `NativeERC20Fetcher` component. GasAdjuster is *always* `Some`, but fetcher is `None`, no updates running for it. Updates for GasAdjuster are running depending on whether `get_or_init` has been called on the corresponding singleton.
- `NativeERC20Fetcher` is provided, and we have at least one `gas_adjuster.get_or_init()`. GasAdjuster is `Some` and now fetcher is also `Some`, so we will run the tasks both for GasAdjuster and fetcher.
- ❗ `NativeERC20Fetcher` is provided, but we never call `gas_adjuster.get_or_init()` (this may happen if a combination of components provided does not require access to L1). Then `GasAdjuster` task will not run, but Fetcher task will run because we've instantiated the object already.

So note the pattern:
- If there is no `NativeERC20Fetcher` component, no task will run.
- If there is `NativeERC20Fetcher` component, the task will *always* run.

Which doesn't correspond to the intended singleton logic.

Reply:
So the fix is then to remove this line where we instantiate the fetcher if the component has been selected

```
let native_erc20_fetcher = if let Some(fetcher_singleton) = &mut fetcher_component {
    let fetcher = fetcher_singleton
        .get_or_init()
        .await
        .context("fetcher.get_or_init()")?;
        Some(fetcher)
    } else {
        None
};
```

and instead do it when initiating the `GasAdjuster`, that way we are only initiating the Fetcher singleton if there's someone that uses it (in this case the `GasAdjuster`). Is my understanding correct?

Reply:
That's one way, but it may be hard to implement (because you'll have to do it in any place where `get_or_init` is called on `GasAdjuster` singleton).
Another way would be to add the `update` method to the `ConversionRateFetcher` method and call it from `GasAdjuster::run` (e.g. `GasAdjuster::run` will be responsible for keeping both itself and `ConversionRateFetcher` up to date).

Reply:
By `update` method here do you mean a method on the `NativeTokenFetcher` that updates the stored conversion rate value by making the request to the API? In that case, if the `GasAdjuster` is the one responsible of calling `update` why have a separate task for the fetcher?

Reply:
Yep, method to update `NativeTokenFetcher` (or rather, any implementation of that trait).
And yep, with such an approach no dedicated task for it is needed, it'll be handled by the gas adjuster task.

Reply:
Made a PR with the fix [here](https://github.com/matter-labs/zksync-era/pull/1360)